File: app/src/main/java/com/checkin/app/checkin/Search/SearchActivity.java
package com.checkin.app.checkin.Search;

import android.arch.lifecycle.ViewModelProviders;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.design.widget.TabLayout;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentPagerAdapter;
import android.support.v4.view.ViewPager;
import android.support.v7.app.AppCompatActivity;
import android.view.View;

import com.checkin.app.checkin.R;
import com.checkin.app.checkin.Shop.ShopPublicProfile.ShopActivity;
import com.checkin.app.checkin.User.NonPersonalProfile.UserProfileActivity;
import com.miguelcatalan.materialsearchview.MaterialSearchView;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

public class SearchActivity extends AppCompatActivity implements SearchResultInteraction {
    private static final String TAG = SearchActivity.class.getSimpleName();

    public static final String KEY_RESULT_PK = "result.pk";
    public static final String KEY_RESULT_NAME = "result.name";
    public static final String KEY_RESULT_IMAGE = "result.image";
    public static final String KEY_RESULT_TYPE = "result.type";

    public static final String KEY_SEARCH_MODE = "search.mode";
    public static final int MODE_SELECT = 1;
    public static final int MODE_SEARCH = 2;

    public static final String KEY_SEARCH_TYPE = "search.type";
    public static final int TYPE_ALL = 0;
    public static final int TYPE_PEOPLE = 1;
    public static final int TYPE_RESTAURANT = 2;

    // TODO: Apply search filters!
    public static final String KEY_SEARCH_FILTER = "search.filter";
    public static final int FILTER_ALL = 0;
    public static final int FILTER_CONNECTED = 1;
    public static final int FILTER_NOT_CONNECTED = 2;

    @BindView(R.id.pager_search_type) ViewPager vPagerSearchType;
    @BindView(R.id.search_view) MaterialSearchView vSearch;
    @BindView(R.id.tabs_search) TabLayout vTabs;

    private ResultTypePagerAdapter mResultTypeAdapter;
    private SearchViewModel mViewModel;
    private SuggestionsFragment mSuggestionsFragment;

    private int mSearchMode = MODE_SEARCH;
    private int mSearchType = TYPE_ALL;
    private int mSearchFilter = FILTER_ALL;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_search);

        ButterKnife.bind(this);

        mSearchMode = getIntent().getIntExtra(KEY_SEARCH_MODE, MODE_SEARCH);
        mSearchType = getIntent().getIntExtra(KEY_SEARCH_TYPE, TYPE_ALL);
        mSearchFilter = getIntent().getIntExtra(KEY_SEARCH_FILTER, FILTER_ALL);

        mViewModel = ViewModelProviders.of(this).get(SearchViewModel.class);

        setupSearch();
        setupUi();
    }

    private void setupUi() {
        if (shouldShowSuggestions()) {
            mSuggestionsFragment = SuggestionsFragment.newInstance();
            showSuggestions();
        }
        else {
            vSearch.showSearch(false);
        }
        setupSearchTabs();
        vPagerSearchType.setAdapter(mResultTypeAdapter);
        vTabs.setupWithViewPager(vPagerSearchType);
    }

    private void setupSearchTabs() {
        mResultTypeAdapter = new ResultTypePagerAdapter(getSupportFragmentManager());

        PeopleResultFragment fragmentPeopleResult = null;
        RestaurantResultFragment fragmentRestaurantResult = null;
        AllResultFragment fragmentAllResult = null;

        if (showPeopleResults()) {
            fragmentPeopleResult = PeopleResultFragment.newInstance(this);
        } else if (showRestaurantResults()) {
            fragmentRestaurantResult = RestaurantResultFragment.newInstance(this);
        } else if (showAllResults()) {
            fragmentPeopleResult = PeopleResultFragment.newInstance(this);
            fragmentRestaurantResult = RestaurantResultFragment.newInstance(this);
            fragmentAllResult = AllResultFragment.newInstance(this);
            showTabs();
        }

        if (fragmentAllResult != null)
            mResultTypeAdapter.addFragment(fragmentAllResult, "All");
        if (fragmentPeopleResult != null)
            mResultTypeAdapter.addFragment(fragmentPeopleResult, "People");
        if (fragmentRestaurantResult != null)
            mResultTypeAdapter.addFragment(fragmentRestaurantResult, "Restaurant");
    }

    private void setupSearch() {
        vSearch.setStartFromRight(false);
        vSearch.setBackIcon(getDrawable(R.drawable.ic_appbar_back));
        vSearch.setOnQueryTextListener(new MaterialSearchView.OnQueryTextListener() {
            @Override
            public boolean onQueryTextSubmit(String query) {
                mViewModel.getSearchResults(query);
                return true;
            }

            @Override
            public boolean onQueryTextChange(String newText) {
                mViewModel.getSearchResults(newText);
                return true;
            }

            @Override
            public boolean onQueryClear() {
                return false;
            }
        });

        vSearch.setOnSearchViewListener(new MaterialSearchView.SearchViewListener() {
            @Override
            public void onSearchViewShown() {
            }

            @Override
            public void onSearchViewClosed() {
                resetUi();
            }
        });
    }

    private void showSuggestions() {
        getSupportFragmentManager().beginTransaction()
                .add(R.id.container_search_suggestions, mSuggestionsFragment)
                .commit();
    }

    private void removeSuggestions() {
        getSupportFragmentManager().beginTransaction()
                .remove(mSuggestionsFragment)
                .commit();
    }

    private void resetUi() {
        if (shouldShowSuggestions()) {
            showSuggestions();
        } else {
            onCancelSearch();
        }
    }

    @OnClick(R.id.tv_search)
    public void onClickSearch(View v) {
        if (shouldShowSuggestions()) {
            removeSuggestions();
        }
        vSearch.showSearch(true);
    }

    @OnClick(R.id.btn_back)
    public void onClickBack(View v) {
        onBackPressed();
    }

    private void showTabs() {
        vTabs.setVisibility(View.VISIBLE);
    }

    private boolean shouldShowSuggestions() {
        return mSearchMode == MODE_SEARCH;
    }

    private boolean shouldReturnResult() {
        return mSearchMode == MODE_SELECT;
    }

    private boolean showAllResults() {
        return mSearchType == TYPE_ALL;
    }

    private boolean showPeopleResults() {
        return mSearchType == TYPE_PEOPLE;
    }

    private boolean showRestaurantResults() {
        return mSearchType == TYPE_RESTAURANT;
    }

    @Override
    public void onBackPressed() {
        if (vSearch.isSearchOpen()) {
            vSearch.closeSearch();
        } else if (shouldReturnResult()) {
            onCancelSearch();
        } else {
            super.onBackPressed();
        }
    }

    private void onCancelSearch() {
        setResult(RESULT_CANCELED);
        finish();
    }

    @Override
    public void onClickResult(SearchResultModel searchItem) {
        if (shouldReturnResult()) {
            Intent data = new Intent()
                    .putExtra(KEY_RESULT_NAME, searchItem.getName())
                    .putExtra(KEY_RESULT_IMAGE, searchItem.getImageUrl())
                    .putExtra(KEY_RESULT_PK, searchItem.getPk())
                    .putExtra(KEY_RESULT_TYPE, searchItem.getType().type);
            setResult(RESULT_OK, data);
            finish();
        } else {
            showInfo(searchItem);
        }
    }

    @Override
    public boolean onLongClickResult(SearchResultModel searchItem) {
        if (shouldReturnResult()) {
            showInfo(searchItem);
            return true;
        }
        return false;
    }

    private void showInfo(SearchResultModel searchItem) {
        if (searchItem.isTypeRestaurant())
            showRestaurantInfo(searchItem.getPk());
        else if (searchItem.isTypePeople())
            showUserInfo(searchItem.getPk());
    }

    private void showUserInfo(String pk) {
        Intent intent = new Intent(this, UserProfileActivity.class);
        intent.putExtra(UserProfileActivity.KEY_PROFILE_USER_ID, Long.valueOf(pk));
        startActivity(intent);
    }

    private void showRestaurantInfo(String pk) {
        Intent intent = new Intent(this, ShopActivity.class);
        intent.putExtra(ShopActivity.KEY_SHOP_PK, pk);
        startActivity(intent);
    }

    private static class ResultTypePagerAdapter extends FragmentPagerAdapter {
        private final List<Fragment> mFragments = new ArrayList<>();
        private final List<String> mFragmentTitles = new ArrayList<>();

        ResultTypePagerAdapter(FragmentManager fm) {
            super(fm);
        }

        void addFragment(Fragment fragment, String title) {
            mFragments.add(fragment);
            mFragmentTitles.add(title);
        }

        @Nullable
        @Override
        public CharSequence getPageTitle(int position) {
            return mFragmentTitles.get(position);
        }

        @Override
        public Fragment getItem(int position) {
            return mFragments.get(position);
        }

        @Override
        public int getCount() {
            return mFragments.size();
        }
    }
}

File: app/src/main/java/com/checkin/app/checkin/Search/SearchActivity.kt
package com.checkin.app.checkin.Search

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.View
import androidx.activity.viewModels
import androidx.annotation.IntDef
import androidx.appcompat.widget.Toolbar
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.fragment.app.FragmentPagerAdapter
import androidx.viewpager.widget.ViewPager
import butterknife.BindView
import butterknife.ButterKnife
import com.checkin.app.checkin.R
import com.checkin.app.checkin.Utility.Utils
import com.checkin.app.checkin.misc.activities.BaseActivity
import com.checkin.app.checkin.restaurant.activities.openPublicRestaurantProfile
import com.google.android.material.tabs.TabLayout
import com.miguelcatalan.materialsearchview.MaterialSearchView
import com.miguelcatalan.materialsearchview.MaterialSearchView.SearchViewListener
import java.util.*

class SearchActivity : BaseActivity(), SearchResultInteraction {
    @BindView(R.id.pager_search_type)
    internal lateinit var vPagerSearchType: ViewPager
    @BindView(R.id.search_view)
    internal lateinit var vSearch: MaterialSearchView
    @BindView(R.id.tabs_search)
    internal lateinit var vTabs: TabLayout
    @BindView(R.id.toolbar_search)
    internal lateinit var toolbar: Toolbar

    private val mResultTypeAdapter: ResultTypePagerAdapter by lazy {
        ResultTypePagerAdapter(supportFragmentManager)
    }
    private val mViewModel: SearchViewModel by viewModels()
    private val mSearchMode by lazy {
        intent.getIntExtra(KEY_SEARCH_MODE, MODE_SEARCH)
    }
    private val mSearchType by lazy {
        intent.getIntExtra(KEY_SEARCH_TYPE, TYPE_ALL)
    }
    private val mSearchFilter by lazy {
        intent.getIntExtra(KEY_SEARCH_FILTER, FILTER_ALL)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_search)
        ButterKnife.bind(this)

        setSupportActionBar(toolbar)
        supportActionBar?.run {
            title = ""
            setDisplayHomeAsUpEnabled(true)
        }
        when (mSearchType) {
            TYPE_PEOPLE -> mViewModel.setup(true, false)
            TYPE_RESTAURANT -> mViewModel.setup(false, true)
            else -> mViewModel.setup(true, true)
        }
        setupSearch()
        setupUi()
    }

    private fun setupUi() {
        vSearch.showSearch(false)
        setupSearchTabs()
        vPagerSearchType.adapter = mResultTypeAdapter
        vTabs.setupWithViewPager(vPagerSearchType)
    }

    private fun setupSearchTabs() {
        var fragmentPeopleResult: PeopleResultFragment? = null
        var fragmentRestaurantResult: RestaurantResultFragment? = null
        var fragmentAllResult: AllResultFragment? = null
        when {
            showPeopleResults() -> {
                fragmentPeopleResult = PeopleResultFragment.newInstance(this)
            }
            showRestaurantResults() -> {
                fragmentRestaurantResult = RestaurantResultFragment.newInstance(this)
            }
            showAllResults() -> {
                fragmentPeopleResult = PeopleResultFragment.newInstance(this)
                fragmentRestaurantResult = RestaurantResultFragment.newInstance(this)
                fragmentAllResult = AllResultFragment.newInstance(this)
                showTabs()
            }
        }
        if (fragmentAllResult != null) mResultTypeAdapter.addFragment(fragmentAllResult, "All")
        if (fragmentPeopleResult != null) mResultTypeAdapter.addFragment(fragmentPeopleResult, "People")
        if (fragmentRestaurantResult != null) mResultTypeAdapter.addFragment(fragmentRestaurantResult, "Restaurant")
        mResultTypeAdapter.notifyDataSetChanged()
    }

    private fun setupSearch() {
        vSearch.setStartFromRight(false)
        vSearch.setBackIconGone()
        vSearch.setOnQueryTextListener(object : MaterialSearchView.OnQueryTextListener {
            override fun onQueryTextSubmit(query: String): Boolean {
                mViewModel.getSearchResults(query)
                return true
            }

            override fun onQueryTextChange(newText: String): Boolean {
                mViewModel.getSearchResults(newText)
                return true
            }

            override fun onQueryClear(): Boolean {
                mViewModel.resetResults()
                return true
            }
        })
        vSearch.setOnSearchViewListener(object : SearchViewListener {
            override fun onSearchViewShown() {}
            override fun onSearchViewClosed() {
                resetUi()
            }
        })
    }

    private fun resetUi() {
        onCancelSearch()
    }

    override fun onSupportNavigateUp(): Boolean {
        finish()
        return true
    }

    private fun showTabs() {
        vTabs.visibility = View.VISIBLE
    }

    private fun shouldReturnResult(): Boolean {
        return mSearchMode == MODE_SELECT
    }

    private fun showAllResults(): Boolean {
        return mSearchType == TYPE_ALL
    }

    private fun showPeopleResults(): Boolean {
        return mSearchType == TYPE_PEOPLE
    }

    private fun showRestaurantResults(): Boolean {
        return mSearchType == TYPE_RESTAURANT
    }

    override fun onBackPressed() {
        when {
            vSearch.isSearchOpen -> {
                vSearch.closeSearch()
            }
            shouldReturnResult() -> {
                onCancelSearch()
            }
            else -> {
                super.onBackPressed()
            }
        }
    }

    private fun onCancelSearch() {
        setResult(Activity.RESULT_CANCELED)
        finish()
    }

    override fun onClickResult(searchItem: SearchResultModel) {
        if (shouldReturnResult()) {
            val data = Intent()
                    .putExtra(KEY_RESULT_NAME, searchItem.name)
                    .putExtra(KEY_RESULT_IMAGE, searchItem.imageUrl)
                    .putExtra(KEY_RESULT_PK, searchItem.pk)
                    .putExtra(KEY_RESULT_TYPE, searchItem.type)
            setResult(Activity.RESULT_OK, data)
            finish()
        } else {
            showInfo(searchItem)
        }
    }

    override fun onLongClickResult(searchItem: SearchResultModel): Boolean {
        if (shouldReturnResult()) {
            showInfo(searchItem)
            return true
        }
        return false
    }

    override fun onClickFollowResult(searchItem: SearchResultModel) {
        Utils.toast(this, "Unsupported operation.")
    }

    private fun showInfo(searchItem: SearchResultModel) {
        if (searchItem.isTypeRestaurant) showRestaurantInfo(searchItem.pk)
        else if (searchItem.isTypePeople) showUserInfo(searchItem.pk)
    }

    private fun showUserInfo(pk: Long) {}
    private fun showRestaurantInfo(pk: Long) {
        openPublicRestaurantProfile(pk)
    }

    private class ResultTypePagerAdapter internal constructor(fm: FragmentManager?) : FragmentPagerAdapter(fm!!) {
        private val mFragments: MutableList<Fragment> = ArrayList()
        private val mFragmentTitles: MutableList<String> = ArrayList()
        fun addFragment(fragment: Fragment, title: String) {
            mFragments.add(fragment)
            mFragmentTitles.add(title)
        }

        override fun getPageTitle(position: Int): CharSequence? {
            return mFragmentTitles[position]
        }

        override fun getItem(position: Int): Fragment {
            return mFragments[position]
        }

        override fun getCount(): Int {
            return mFragments.size
        }
    }

    companion object {
        const val KEY_RESULT_PK = "result.pk"
        const val KEY_RESULT_NAME = "result.name"
        const val KEY_RESULT_IMAGE = "result.image"
        const val KEY_RESULT_TYPE = "result.type"
        const val KEY_SEARCH_MODE = "search.mode"
        const val MODE_SELECT = 1
        const val MODE_SEARCH = 2
        const val KEY_SEARCH_TYPE = "search.type"
        const val TYPE_ALL = 0
        const val TYPE_PEOPLE = 1
        const val TYPE_RESTAURANT = 2
        // TODO: Apply search filters!
        const val KEY_SEARCH_FILTER = "search.filter"
        const val FILTER_ALL = 0
        const val FILTER_CONNECTED = 1
        const val FILTER_NOT_CONNECTED = 2

        @IntDef(TYPE_ALL, TYPE_PEOPLE, TYPE_RESTAURANT)
        @Retention(AnnotationRetention.SOURCE)
        annotation class SearchType

        @IntDef(MODE_SEARCH, MODE_SELECT)
        @Retention(AnnotationRetention.SOURCE)
        annotation class SearchMode

        fun withIntent(context: Context, @SearchType searchType: Int = TYPE_ALL, @SearchMode searchMode: Int = MODE_SEARCH) = Intent(context, SearchActivity::class.java).apply {
            putExtra(KEY_SEARCH_TYPE, searchType)
            putExtra(KEY_SEARCH_MODE, searchMode)
        }
    }
}